File: Data Structures & Algorithms/merge-intervals/submission-2.py
# class Solution:
#     def merge(self, intervals: List[List[int]]) -> List[List[int]]:
#         res = []
#         intervals.sort(key=lambda x: x[0])
#         res.append(intervals[0])
#         for i in range(1, len(intervals)):  
#             cur = intervals[i]
#             last = res[-1]
#             if cur[0] <= last[1]:
#                 last[1] = max(cur[1], last[1])
#             else:
#                 res.append(cur)
        
#         return res

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        res = []
        intervals.sort(key=lambda x: x[0])
        
        for interval in intervals:
            if not res or res[-1][1] < interval[0]:
                res.append(interval)
            else:
                res[-1][1] = max(interval[1], res[-1][1])
        
        return res

File: Data Structures & Algorithms/maximum-product-subarray/submission-2.java
class Solution {
    public int maxProduct(int[] nums) {
        int n = nums.length;
        if (n == 0) return 0;
        int maxNum = nums[0];
        int minNum = nums[0];
        int res = nums[0];
        for (int i = 1; i < n; i++) {
            int cur = nums[i];
            int tempMax = maxNum;
            int tempMin = minNum;
            maxNum = Math.max(Math.max(cur, tempMax * cur), tempMin * cur);
            minNum = Math.min(Math.min(cur, tempMin * cur), tempMax * cur);
            res = Math.max(res, maxNum);
        }
        return res;
    }
}
